File: array/74.py
#这个题面试时遇到过, 本身matrix是有特点的, 如果从左下角开始搜索 就可以看到规律
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        i = len(matrix) - 1
        j = 0
        flag = 0
        if not matrix or not matrix[0]:
            return False
        while i >= 0 and j < len(matrix[0]):
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] < target:
                j += 1
            else:
                i -= 1
        return False
            

File: binary_search/74.cc
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty())
            return false;
        int j = 0;
        while (j < matrix.size()) {
            if ((j+1) < matrix.size() && target > matrix[j+1][0]) {
                j++;
            } else {
                // use binary search
                if (search(matrix[j], target))
                    return true;
                j++;
            }
        }
        return false;
    }
private:
    bool search(std::vector<int>& nums, int target) {
        int start = 0, end = nums.size()-1;
        int mid = (start + end) / 2;
        while (start <= end) {
            if (target == nums[mid]) {
                return true;
            } else if (target > nums[mid]) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
            mid = (start + end) / 2;
        }
        return false;
    }
};
